src/helpers.ml

```diff
@@ -33,3 +35,61 @@ let trim_comments comment =
       with Stdlib.Not_found -> comment
   in
   aux comment 0 false
+
+let action_with_new_installation_token ~bot_info ~key ~app_id ~owner ~repo
+    ~(installation_tokens : (string, string * float) Base.Hashtbl.t) action =
+  GitHub_app.get_installation_token ~bot_info ~key ~app_id ~owner ~repo
+  >>= function
+  | Ok (github_token, expiration_date) ->
+      let _ =
+        Hashtbl.add installation_tokens ~key:owner
+          ~data:(github_token, expiration_date)
+      in
+      let bot_info : Bot_info.t = {bot_info with github_token} in
+      action ~bot_info
+  | Error _ ->
+      action ~bot_info
+
+let action_as_github_app ~bot_info ~key ~app_id ~owner ~repo
+    ~installation_tokens action =
+  (* Executes an action with an installation token if the repository has
+     the GitHub app installed.
+     Generates a new installation token if the existing one has expired. *)
+  match Hashtbl.find installation_tokens owner with
+  | Some (github_token, expiration_date) ->
+      if Float.(expiration_date < Unix.time ()) then (
```

Comment:
I thought you wanted to take a 20 minutes margin. You removed it from `GitHub_app` without reintroducing it here.

Reply:
The 20 minutes margin is still in `GitHub_app,` no? The part where we set the expiration date.

Reply:
Oh right, I misread.